### ISSUE TITLE
rails console on production uses spring

### ISSUE THREAD

**Issue by visibletrap** [rails/spring#318]
I have been using spring since rails 4.1 released and I really like it. Yesterday we found that there's spring processes running on our production servers. We traced down and found that spring was started when we used rails console with command `bundle exec rails console production`. We don't want it to be used on production.

Regards your comment on this at https://github.com/rails/rails/pull/12958#issuecomment-28915116

> It's already handled by the fact that spring is in the development group of the gemfile. So it's not installed on the server. If it's not installed, the binstub just bypasses spring and runs the underlying command.

What we have experienced is different, spring is installed on our production servers even though it's in :development group.

I'm curious, do you expect us to run `bundle install` with `--without development` on production servers? We haven't had need for `--without development` parameter. We have been using only `bundle install --deployment` so far.

Maybe we should update the documentation for this?

**Comment by amw** [comment]
Also, I sometimes run `rails c production` or `rails c staging` on my dev machine (which has spring installed) and I would not want these processes to use spring. I would like them to run clean, as in production, reloading config files each time they start.

**Comment by squaresurf** [comment]
Spring should listen to the gem groups before loading. Requiring `bundle install` to be run in a non standard fashion is not a fix in my opinion.

I had to add `DISABLE_SPRING=1` to my production environment to fix this which feels more like a hack than a fix.

**Comment by lendyme** [comment]
I've had to temporarily patch `bin/rails` to get `rails console` to work on Heroku. Setting DISABLE_SPRING in Heroku ENV has no effect.

    #!/usr/bin/env ruby
    unless ENV["DISABLE_SPRING"]
      begin
        spring_bin_path = File.expand_path('../spring', __FILE__)
        load spring_bin_path
      rescue LoadError => e
        raise unless e.message.end_with? spring_bin_path, 'spring/binstub'
      end
    end
    APP_PATH = File.expand_path('../../config/application',  __FILE__)
    require_relative '../config/boot'
    require 'rails/commands'

**Comment by sintro** [comment]
> Spring should listen to the gem groups before loading. Requiring bundle install to be run in a non standard fashion is not a fix in my opinion.
I had to add DISABLE_SPRING=1 to my production environment to fix this which feels more like a hack than a fix.

Absolutely agree. I always installed gems on deployment server with just `bundle install`, even without --deployment parameter. It is not feels good for me now to know that I must always remember to include thing ugly '--without test development' to bundle install only to prevent one gem from being installed. 
Creating some aliases for this command is another hack. Is it really hard to code the feature of checking if the environment is production and prevent spring starting?

**Comment by nitzanav** [comment]
I added `ENV['RAILS_ENV'] == 'production'` condition which works in the case of`RAILS_ENV=production rails c` but not in the case of `rails c production`

```
#!/usr/bin/env ruby
unless ENV["DISABLE_SPRING"] || ENV['RAILS_ENV'] == 'production'
    begin
    load File.expand_path('../spring', __FILE__)
  rescue LoadError => e
    raise unless e.message.include?('spring')
  end
end
APP_PATH = File.expand_path('../../config/application', __FILE__)
require_relative '../config/boot'
require 'rails/commands'
```

**Comment by ahoward** [comment]
yes - relying on not having the gem installed is *bad* form.  it's easy to boot a console on a production machine.

**Comment by black-snow** [comment]
Is this still the default behaviour?

**Comment by agis** [comment]
@jonleighton @jeremy @rafaelfranca How about https://github.com/agis/spring/commit/4519522d422e7e98973c89662236e22dc384d4a5? If that looks good I could submit a PR.

**Comment by jonleighton** [comment]
👍

**Comment by mech** [comment]
Still a problem even when specifying this in my Dockerfile, the Spring is still loaded:

```Dockerfile
ENV BUNDLER_WITHOUT development test
RUN setuser app bundle install --jobs 4 --retry 3
```

Have not idea how to make it go away. The Spring gem is indeed on my development group btw.
